pkg/virt-launcher/virtwrap/cmd-server/client/client.go

```diff
@@ -0,0 +1,238 @@
+	Ping() error
+	Close()
+}
+
+type VirtLauncherClient struct {
+	client *rpc.Client
+}
+
+func ListAllSockets(baseDir string) ([]string, error) {
+	var socketFiles []string
+
+	fileDir := filepath.Join(baseDir, "sockets")
+	exists, err := diskutils.FileExists(fileDir)
+	if err != nil {
+		return nil, err
+	}
+
+	if exists == false {
+		return socketFiles, nil
+	}
+
+	files, err := ioutil.ReadDir(fileDir)
+	if err != nil {
+		return nil, err
+	}
+	for _, file := range files {
+		socketFiles = append(socketFiles, filepath.Join(fileDir, file.Name()))
+	}
+	return socketFiles, nil
+}
+
+func SocketsDirectory(baseDir string) string {
+	return filepath.Join(baseDir, "sockets")
+}
+
+func SocketFromNamespaceName(baseDir string, namespace string, name string) string {
+	sockFile := namespace + "_" + name + "_sock"
+	return filepath.Join(SocketsDirectory(baseDir), sockFile)
+}
+
+func GetClient(socketPath string) (LauncherClient, error) {
+	conn, err := rpc.Dial("unix", socketPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &VirtLauncherClient{client: conn}, nil
+}
+
+func (c *VirtLauncherClient) Close() {
+	c.client.Close()
+}
+
+func (c *VirtLauncherClient) genericSendCmd(args *Args, cmd string) (*Reply, error) {
+	reply := &Reply{}
+
+	err := c.client.Call(cmd, args, reply)
+	if IsDisconnected(err) {
+		return reply, err
+	} else if err != nil {
+		msg := fmt.Sprintf("unknown error encountered sending command %s: %s", cmd, err.Error())
+		return reply, fmt.Errorf(msg)
+	} else if reply.Success != true {
+		msg := fmt.Sprintf("server error. command %s failed: %s", cmd, reply.Message)
+		return reply, fmt.Errorf(msg)
+	}
+	return reply, nil
+}
+
+func (c *VirtLauncherClient) ShutdownVirtualMachine(vm *v1.VirtualMachine) error {
+	cmd := "Launcher.Shutdown"
+
+	vmJSON, err := json.Marshal(*vm)
+	if err != nil {
+		return err
+	}
+	args := &Args{
+		VMJSON: string(vmJSON),
+	}
+	_, err = c.genericSendCmd(args, cmd)
+
+	return err
+}
+
+func (c *VirtLauncherClient) KillVirtualMachine(vm *v1.VirtualMachine) error {
+	cmd := "Launcher.Kill"
+
+	vmJSON, err := json.Marshal(*vm)
+	if err != nil {
+		return err
+	}
+	args := &Args{
+		VMJSON: string(vmJSON),
+	}
+	_, err = c.genericSendCmd(args, cmd)
+
+	return err
+}
+
+func (c *VirtLauncherClient) ListDomains() ([]*api.Domain, error) {
+	var list []*api.Domain
+	cmd := "Launcher.ListDomains"
+
+	args := &Args{}
+
+	reply, err := c.genericSendCmd(args, cmd)
+	if err != nil {
+		return list, err
+	}
+
+	err = json.Unmarshal([]byte(reply.DomainListJSON), &list)
+
+	if err != nil {
+		return list, err
+	}
+	return list, nil
+
+}
+func (c *VirtLauncherClient) StartVirtualMachine(vm *v1.VirtualMachine, secrets map[string]*k8sv1.Secret) error {
+
+	cmd := "Launcher.Start"
+
+	vmJSON, err := json.Marshal(*vm)
+	if err != nil {
+		return err
+	}
+	secretJSON, err := json.Marshal(secrets)
+	if err != nil {
+		return err
+	}
+
+	args := &Args{
+		VMJSON:          string(vmJSON),
```

Comment:
I would feel better if we understand the  issue we are facing, which we are working around by using json.

Reply:
```
type ResourceList map[ResourceName]resource.Quantity
```

Anything that uses that map (like memory or cpu requests/limits) gets lost if we don't serialize/deserialize the VirtualMachine as a JSON object.

Reply:
Could you give me a before RPC and after RPC example?

Reply:
before
```
apiVersion: kubevirt.io/v1alpha1
kind: VirtualMachine
metadata:
  name: testvm
spec:   
  terminationGracePeriodSeconds: 0                                  
  domain:
    resources:
      requests:
        memory: 64M
```

after
```apiVersion: kubevirt.io/v1alpha1
kind: VirtualMachine
metadata:
  name: testvm
spec:   
  terminationGracePeriodSeconds: 0                                  
  domain:
    resources:
      requests:
        memory:
```

the memory quantity gets lost

Reply:
ok, so the reason is that the Quantity type has hidden fields, and the serializer by default only takes public fields. Therefore it keeps the format, but not the actual value ... A custom MarshalBinary() might solve that. However not sure if it is easy to do ...

Reply:
ok, so I think I have a solution. Could you implement "MarshalBinary()" and "UnmarshalBinary()" on our VM type? In there just marshal the whole struct into json and convert it to bytes. Then you can remove all custom json marshalling actions regarding to RPC. That would give me a better feeling.

Reply:
I think on the long run, we should use protopuf for rpc, since k8s takes care that their types work with protobuf (and aparently don't care about normal rpc).

Reply:
> Could you implement "MarshalBinary()" and "UnmarshalBinary()" on our VM type

that seems a little overkill. Is there a simple way of doing this?  The binary encoding that golang provides is the same thing that rpc uses, which stripped the memory requests.

Reply:
i see what you're getting at now.  We'll just use the a MarshalBinary function and wrap the json encode in it. I'm fine with doing that.

Reply:
fixed